### DISCUSSION TITLE
Changing fuel on fleet

### DISCUSSION THREAD

**Discussion by GunnarQuante** [contrailcirrus/pycontrails#84] [Q&A] [answered]
For some weeks I am trying to change the fuel used on a fleet/list of flights from JetA to a SAFBlend. I have a fleet of e.g., 100 flights and would like to run Cocip on them once with JetA and once with a 100% SAFBlend. When I loop through each flight in the fleet, change the fuel and run Cocip on each flight individually, the overall contrail EF changes. This is not the case if I change the fuel for the entire fleet. In this case, the overal contrail EF of the JetA run to the SAFBlend run is identical. To reduce computational time, I am looking for a way to run Cocip on the entire fleet: 

def CoCiP_by_fuels(met, rad, flights_in, fuel):
    
    # Use BADA3 for performance modelling
    bada_model = BADAFlight(bada_priority=3)
   
    params = {
        "process_emissions": True,
        "verbose_outputs": False,
        "humidity_scaling": ConstantHumidityScaling(rhi_adj=0.98),
    }
    cocip = Cocip(met=met, rad=rad, params=params, aircraft_performance=bada_model)

        Fleet = pycontrails.Fleet.from_seq(flights_in)
        Fleet.fuel = SAFBlend(100)
        print(Fleet.fuel.fuel_name)
        Fleet_out=cocip.eval(Fleet)
        print(Fleet_out.fuel.fuel_name)
return Fleet_out 

Note that also the fuel_name changes from Fleet to Fleet_out. I suppose I am missing something here, but after quite some reading I still can't find a fix.

**Comment by zebengberg** [comment]
Hi @GunnarQuante 

I've been able to reproduce the bug you've found. Thanks for alerting us!

I believe I've fixed the issue in f8d561b4ea5b60bff1170ff650b1e869571ba599 and c660f9c604df3a455c0858f232dd55b3a2a4255d. This will be included when we release pycontrails 0.47 (hopefully later today).

My hope is that the snippet you've included will work as you expect when you upgrade pycontrails. You can also specify the fuel type on each Flight instance up front, and the Fleet will automatically pick it up (so no need to monkey patch the instance after instantiation).

**Reply by GunnarQuante** [reply]
Hi @zebengberg, 

great, thanks a lot! I'll try with v0.47 later on :)

**Reply by GunnarQuante** [reply]
After upgrading to ypcontrails 0.47 there seems to be a problem with transferring the parameters for the bada model. When I run the snippet from above, I now get the following message: 

    bada_model = BADAFlight(bada_priority=3)

  File ~\.conda\envs\traffic\lib\site-packages\pycontrails\core\models.py:182 in __init__
    self._load_params(params, **params_kwargs)

  File ~\.conda\envs\traffic\lib\site-packages\pycontrails\core\models.py:330 in _load_params
    self.params = base_params.as_dict()

TypeError: descriptor 'params' for 'Model' objects doesn't apply to a 'BADAFlight' object

Or am I missing something here?

**Reply by mlshapiro** [reply, marked as answer]
Hi @GunnarQuante - I will let @zebengberg confirm here, but the 0.47 release had a companion update in `pycontrails-bada`. Try pulling / installing the latest changes from here https://github.com/contrailcirrus/pycontrails-bada

```
pip install -U "pycontrails-bada @ git+ssh://git@github.com/contrailcirrus/pycontrails-bada.git"
```

**Reply by GunnarQuante** [reply]
That worked - thanks a lot!
